dynamiqs/core/abstract_solver.py

```diff
@@ -21,15 +23,31 @@ def run(self) -> PyTree:
 class BaseSolver(AbstractSolver):
     ts: Array
     y0: Array
-    H: TimeArray
+    H: Array | TimeArray
     Es: Array
     solver: Solver
     gradient: Gradient | None
     options: Options
 
+    def save(self, y: Array) -> dict[str, Array]:
+        saved = {}
+        if self.options.save_states:
+            saved['ysave'] = y
+        if self.Es is not None and len(self.Es) > 0:
+            saved['Esave'] = bexpect(self.Es, y)
+        return saved
+
+    def result(self, saved: dict[str, Array]) -> Result:
+        ysave = saved.get('ysave', None)
+        Esave = saved.get('Esave', None)
+        if Esave is not None:
+            Esave = Esave.swapaxes(-1, -2)
+
+        return Result(self.ts, self.solver, self.gradient, self.options, ysave, Esave)
+
 
 SESolver = BaseSolver
 
 
 class MESolver(BaseSolver):
-    Ls: list[TimeArray]
+    Ls: list[Array | TimeArray]
```

Comment:
Why not `ArrayLike`?

Reply:
This is not the API layer anymore so we type everything with fully converted types in the backend.